### PR TITLE
8319872: AArch64: [vectorapi] Implementation of unsigned (zero extended) casts

### PR DESCRIPTION
**Pull request by e1iu**
Vector API defines zero-extend operations [1], which are going to be intrinsified and generated to `VectorUCastNode` by C2. This patch adds backend implementation for `VectorUCastNode` on AArch64.

The micro benchmark shows significant performance improvement. In my test machine (SVE, 256-bit), the result is shown as below:

```

  Benchmark                     Before     After       Units   Gain
  VectorZeroExtend.byte2Int     3168.251   243012.399  ops/ms  75.70
  VectorZeroExtend.byte2Long    3212.201   216291.588  ops/ms  66.33
  VectorZeroExtend.byte2Short   3391.968   182655.365  ops/ms  52.85
  VectorZeroExtend.int2Long     1012.197    80448.553  ops/ms  78.48
  VectorZeroExtend.short2Int    1812.471   153416.828  ops/ms  83.65
  VectorZeroExtend.short2Long   1788.382   129794.814  ops/ms  71.58
```

On other Neon systems, we can get similar performance boost as a result of intrinsification success.

Since `VectorUCastNode` only used in Vector API's zero extension currently, this patch also adds assertion on nodes' definitions to clarify their usages.

[TEST]
compiler/vectorapi and jdk/incubator/vector passed on NEON and SVE machines.

[1] https://github.com/openjdk/jdk/blob/master/src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorOperators.java#L726

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319872](https://bugs.openjdk.org/browse/JDK-8319872): AArch64: [vectorapi] Implementation of unsigned (zero extended) casts (**Enhancement** - P4)

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)
 * [Xiaohong Gong](https://openjdk.org/census#xgong) (@XiaohongGong - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16670/head:pull/16670` \
`$ git checkout pull/16670`

Update a local copy of the PR: \
`$ git checkout pull/16670` \
`$ git pull https://git.openjdk.org/jdk.git pull/16670/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16670`

View PR using the GUI difftool: \
`$ git pr show -t 16670`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16670.diff">https://git.openjdk.org/jdk/pull/16670.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16670#issuecomment-1811965423)

### CODE REVIEW

File: src/hotspot/cpu/aarch64/aarch64_vector_ad.m4 [hunk 1]
```diff
@@ -2320,6 +2320,69 @@ instruct reinterpret_resize_gt128b(vReg dst, vReg src, pReg ptmp, rFlagsReg cr)
   ins_pipe(pipe_slow);
 %}
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_vector_ad.m4:2322]: The following hunk does not seem to be making good use of the macro processor.

Comment (e1iu) [reply src/hotspot/cpu/aarch64/aarch64_vector_ad.m4:2322]: I updated m4 file. Please help to review, thanks!

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp [hunk 2]
```diff
@@ -1337,29 +1337,33 @@ void C2_MacroAssembler::sve_vmask_lasttrue(Register dst, BasicType bt, PRegister
...
 void C2_MacroAssembler::neon_vector_extend(FloatRegister dst, BasicType dst_bt, unsigned dst_vlen_in_bytes,
-                                           FloatRegister src, BasicType src_bt) {
+                                           FloatRegister src, BasicType src_bt, bool is_unsigned) {
+  xtl_insn ext = is_unsigned ? &C2_MacroAssembler::uxtl : &C2_MacroAssembler::sxtl;
   if (src_bt == T_BYTE) {
     if (dst_bt == T_SHORT) {
       // 4B/8B to 4S/8S
       assert(dst_vlen_in_bytes == 8 || dst_vlen_in_bytes == 16, "unsupported");
-      sxtl(dst, T8H, src, T8B);
+      (this->*ext)(dst, T8H, src, T8B);
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:1352]: One thing might make this cleaner: I suggest you make `_xshll` protected rather than private, then here

`_xshll(is_unsigned, dst, T8H, src, T8B, 0);`

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp [hunk 3]
```diff
@@ -1392,35 +1396,42 @@ void C2_MacroAssembler::neon_vector_narrow(FloatRegister dst, BasicType dst_bt,
...
   if (src_size == B) {
     switch (dst_size) {
     case H:
-      sve_sunpklo(dst, H, src);
+      (this->*unpklo)(dst, H, src);
       break;
     case S:
-      sve_sunpklo(dst, H, src);
-      sve_sunpklo(dst, S, dst);
+      (this->*unpklo)(dst, H, src);
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:1415]: AS above: try making` is_unsigned` a parameter.

Comment (e1iu) [reply src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:1415]: Got it. I will fix it soon. Thanks!

Comment (e1iu) [reply src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp]: compiler/vectorapi and jdk/incubator/vector passed. Full test is running. I would report the result when it has been finished.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 4]
```diff
@@ -3862,19 +3841,35 @@ template<typename R, typename... Rx>
...
+
+  // Signed unpack and extend half of vector - high half
+  void sve_sunpkhi(FloatRegister Zd, SIMD_RegVariant T, FloatRegister Zn) {
+    _sve_xunpk(/* is_unsigned */ false, /* is_high */ true, Zd, T, Zn);
+  }
+
+  // Signed unpack and extend half of vector - low half
+  void sve_sunpklo(FloatRegister Zd, SIMD_RegVariant T, FloatRegister Zn) {
+    _sve_xunpk(/* is_unsigned */ false, /* is_high */ false, Zd, T, Zn);
+  }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3872]: This code expansion does not look right. You should be able to make this change without so much code expansion.

```
#define INSN(NAME, unsigned, high)
void name(FloatRegister Zd, SIMD_RegVariant T, FloatRegister Zn) { \
  _sve_xunpk(unsigned, high, T, Zn);                                            \
}
  INSN(sve_uunpkhi, true, true)
...
```

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp [hunk 5]
```diff
@@ -1393,34 +1392,36 @@ void C2_MacroAssembler::neon_vector_narrow(FloatRegister dst, BasicType dst_bt,
...
+      _sve_xunpk(is_unsigned, /* is_high */ false, dst, S, dst);
       break;
     case D:
-      sve_sunpklo(dst, H, src);
-      sve_sunpklo(dst, S, dst);
-      sve_sunpklo(dst, D, dst);
+      _sve_xunpk(is_unsigned, /* is_high */ false, dst, H, src);
+      _sve_xunpk(is_unsigned, /* is_high */ false, dst, S, dst);
+      _sve_xunpk(is_unsigned, /* is_high */ false, dst, D, dst);
       break;
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:1412]: Why is this change here? It doesn't do anything. Does it?

Comment (e1iu) [reply src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:1412]: `is_unsigned` is also used in this function. It is used in VectorUCastNode for zero extending.

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp (hunk 3 above)
Comment (e1iu) [reply src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp]: Full jtreg passed without new failure.

**Comment by e1iu** [comment]
@theRealAph Could you help to take a look? Thanks.

**Review by theRealAph — APPROVED** [review]
That looks nice. Thanks.

**Review by XiaohongGong — APPROVED** [review]
LGTM!
